### PR TITLE
Reorganize to (incrementally) resemble k8s more closely

### PR DESCRIPTION
**Pull request by krancour**
This PR re-organizes existing code to give the entire repo a more k8s-like structure:

- Moves code around to more closely mirror k8s' directory / package layout and makes the complementary changes to package names, imports, etc.

- Refactors existing code to more closely resemble "the k8s way." This is _mostly_ in the `cmd/controller` package. I drew tremendous inspiration from k8s `cmd/kube-controller-manager`. I basically used it as a template for how `cmd/controller` and its subpackages should look internally.

- Reuses more things from k8s where possible (instead of duplicating).

- Addresses much of #198.
    - This is mostly re-organization / re-factoring of `pkg/controller`.
    - Attempts to reduce overloading of terms like "controller" and "server."

This is an _incremental_ improvement. Don't expect the outcome of this to give us a perfect resemblance to k8s- it's just a big step in that direction.

Note what this PR _doesn't_ do-- it doesn't introduce any functional changes.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -46,30 +46,28 @@ endif
 # "apiserver" instead of "bin/apiserver".
 #########################################################################
 build: .init .generate_files \
-       $(BINDIR)/controller $(BINDIR)/registry $(BINDIR)/k8s-broker \
-       $(BINDIR)/service-catalog $(BINDIR)/user-broker \
-       $(BINDIR)/apiserver
+       $(BINDIR)/controller $(BINDIR)/apiserver \
+       $(BINDIR)/registry $(BINDIR)/k8s-broker $(BINDIR)/user-broker
```
Comment (krancour) [inline Makefile]: Along with `.init .generate_files`, these are the pre-requisite target for the `build` target. `$(BINDIR)/service-catalog` went away because it seemed to be a duplicate of the controller and as for the rest, I just grouped targets related to service catalog components on one line and contrib stuffs on another.

File: cmd/controller/app/options/options.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+			KubeconfigPath: "./kubeconfig",
+		},
+	}
+}
+
+// AddFlags adds flags for a specific CMServer to the specified FlagSet
+func (s *ControllerServer) AddFlags(fs *pflag.FlagSet) {
+	fs.Var(k8scomponentconfig.IPVar{Val: &s.Address}, "address", "The IP address to serve on (set to 0.0.0.0 for all interfaces)")
+	fs.Int32Var(&s.Port, "port", s.Port, "The port that the controller-manager's http service runs on")
+	// fs.StringVar(&s.ConfigPath, "c", ".", "use '-c' option to specify the config file path")
```
Comment (krancour) [inline cmd/controller/app/options/options.go]: I couldn't find where this was even being used. Commenting it was probably experimental on my part. I think I can remove that line altogether.

Comment (krancour) [inline cmd/controller/app/options/options.go]: Double checked and confirmed this wasn't previously used. Instead of commenting it, I am now removing it.

File: Makefile [hunk 3]
```diff
@@ -99,8 +97,8 @@ $(BINDIR)/deepcopy-gen: cmd/libs/go2idl/deepcopy-gen
 	$(DOCKER) glide install --strip-vendor
 	touch $@
 
-.scBuildImage: hack/Dockerfile
-	sed "s/GO_VERSION/$(GO_VERSION)/g" < hack/Dockerfile | \
+.scBuildImage: build/build-image/Dockerfile
+	sed "s/GO_VERSION/$(GO_VERSION)/g" < build/build-image/Dockerfile | \
```
Comment (duglin) [inline Makefile]: Am I correct in assuming that "build" is now the place for "build related artifacts", not "hack" ?  If so, cool.

Is "Dockerfile" the only thing in "build-image"? If so do we need it in a dir?  If you're worried about confusion we could call it "Dockerfile-BuildImage" or something like that.

File: cmd/controller/app/controller.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
+/*
```
Comment (duglin) [inline cmd/controller/app/controller.go]: is this a new file? I just surprised that git doesn't show it as a move with a minor tweak (if anything was changed).

File: pkg/apis/componentconfig/types.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
+/*
```
Comment (duglin) [inline pkg/apis/componentconfig/types.go]: what does "componentconfig" refer to? From the comment at the start of this file I would have expected this to be called "controllerconfig" or something like that.

File: Makefile (hunk 3 above)
Comment (duglin) [inline Makefile]: Should we move other files from hack to build?

Comment (krancour) [inline Makefile]: The name of the game here is re-organizing to incrementally advance our repo in similarity to k8s. This is where they keep the Dockerfile for their development environment, so I took my cues from that.

See path to the Dockerfile in the link below:

https://github.com/kubernetes/kubernetes/blob/master/build/build-image/Dockerfile

**Comment by duglin** [comment]
Overall direction seems good. Would it make sense for this PR to include an update to the docs/DEVGUIDE.md doc that describes the basic layout of our dir structure? So people know what to expect to find under pkg vs contrib vs cmd - as an example.

File: cmd/controller/app/controller.go (hunk 4 above)
Comment (krancour) [inline cmd/controller/app/controller.go]: It's not new, per se, but it's changed enough that git probably doesn't recognize it as the same file. I don't pretend to know git internals to that level.

File: pkg/apis/componentconfig/types.go (hunk 5 above)
Comment (krancour) [inline pkg/apis/componentconfig/types.go]: Good question. Again, I've modeled this after k8s.

This is where k8s defines the types used to encapsulate configuration options for each component (by "component," I mean executable / process). Stuff in `cmd` populates these options from flags.

Yes... I'm blindly following how they do things.

https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/componentconfig/types.go

File: Makefile (hunk 3 above)
Comment (duglin) [inline Makefile]: yea but they have other files in that dir.  A dir for just one file feels really odd to me so I would prefer to wait until we need a "grouping mechanism" before we unnecessarily introduce one.  I would also prefer that we keep all build-related files in one dir, and not split across 'hack' and 'build'.

Comment (krancour) [inline Makefile]: I understand you, but I have to disagree if only for the fact that this is intended to be an _incremental_ step toward looking more like k8s. k8s has both `build` and `hack` dirs, so I imagine we will as well. I have an imperfect understanding of the delineation between those two, but intuitively, I would guess (or hope?) that build-related stuff is isolated to `build` and that much of what is in `hack` are scripts that exist more for the sake of developer convenience? Maybe @pmorie can shed more light on this...

The bottom line (from my own point of view, at least) is that I imagine _more_ stuff moving into the `build` dir over time... we just haven't identified what those things are yet, but in recognition of it, the Dockerfile was preemptively placed where it is, which at least does move us in the direction of looking more like k8s.

File: Makefile (hunk 1 above)
Comment (pmorie) [inline Makefile]: Got you.

File: cmd/controller/app/controller.go (hunk 4 above)
Comment (pmorie) [inline cmd/controller/app/controller.go]: I believe the threshold for distinguishing a move with modification from a delete and an add is that the modification has to touch less than 50% of the file's lines.  I don't have an official citation for this offhand, but there is some kind of similarity threshold.

File: pkg/apis/componentconfig/types.go (hunk 5 above)
Comment (pmorie) [inline pkg/apis/componentconfig/types.go]: The `componentconfig` API in k8s is an API for the configuration options different system components like the API server, kubelet, etc, accept.  @krancour was right to model it in this case, but I expect it will change as we refactor to the kubernetes controller frameworks.

**Comment by krancour** [comment]
@duglin per the difficulties I noted in #217, the bits of refactoring in this PR affecting the `Makefile` weren't fully vetted, because I couldn't even get the build process to work for me from master-- i.e. I had no baseline for comparison...

But after digging into #217, I realized the answer to some of the permissions issues is to put the Docker build contexts somewhere _other_ than `bin` which might not be writable. Looking back to k8s for inspiration, I see more clearly now that much of what is in _their_ `build` dir are build contexts...

So... I amended this PR with e9451611a5abb979dce5658e27629ab8f5faa20e to address #217 _and_ account for all the moving around of Dockerfiles that were part of this reorganization effort. A nice side effect of this is that it obviates the issue you expressed here: https://github.com/kubernetes-incubator/service-catalog/pull/210#discussion_r94601720

If you don't mind giving this another look, with special attention paid to e9451611a5abb979dce5658e27629ab8f5faa20e, that would be great!

**Comment by krancour** [comment]
> can we pull out the Makefile edits for Docker stuff into a new PR?

@duglin maybe...

I didn't initially want to bundle those changes into this PR, but Dockerfiles moved anyway during this refactoring, which means Makefiles had to be updated anyway... and I didn't have a convenient way of vetting those changes because, as we've discussed, your run as non-root strategy doesn't work on Mac + Docker Machine. Fixing the issues I reported in #217 allowed me to assert the rest of these changes are ok.

I can try to undo some of this and test in a Linux VM instead. It's just a question of how much work you want to make for me.

**Comment by krancour** [comment]
@duglin I removed the second commit from the PR, moved my testing to a Linux VM, found a few issues (just broken paths introduced by having moved things around), fixed those and amended the original commit. This all checks out when I build in-container on a Linux VM.

I'll open a separate PR to address #217.

Would like to try and get this merged today. It's a big refactor and I want to avoid myself or anyone else working on big things having to do some tedious rebases.

**Comment by krancour** [comment]
@duglin sorry... I failed to address this comment before:

> Would it make sense for this PR to include an update to the docs/DEVGUIDE.md doc that describes the basic layout of our dir structure?

I like that idea, but the layout of the repo wasn't documented prior to this PR, so I don't think it's 100% incumbent upon this PR to tackle that. At least this PR gets us to a place where the layout, I think, is more consistent and intuitive. I'll gladly open an issue to follow this up with some doc updates and will even assign it to myself.

File: cmd/controller/app/controller.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+	dynClient, err := dynamic.NewClient(kubeconfig)
+	if err != nil {
+		glog.Errorf("Failed to create a dynamic client\n:%v\n", err)
+		return err
+	}
+
+	w, err := watch.NewWatcher(k8sClient, dynClient)
+	if err != nil {
+		panic(fmt.Sprintf("Failed to create a watcher: %v\n", err))
```
Comment (arschles) [inline cmd/controller/app/controller.go]: why panic here and log & return errors above?

File: cmd/controller/app/controller.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+		server := &http.Server{
+			Addr:    net.JoinHostPort(s.Address, strconv.Itoa(int(s.Port))),
+			Handler: mux,
+		}
+		glog.Fatal(server.ListenAndServe())
+	}()
+
+	c.Run()
+
+	panic("unreachable")
```
Comment (arschles) [inline cmd/controller/app/controller.go]: I know you probably didn't touch this code, but can you remove this and return an error instead

File: contrib/pkg/broker/k8s/controller/controller.go [hunk 8]
```diff
@@ -200,34 +199,34 @@ func (c *k8sController) getType(serviceID string, planID string) (*model.Types,
...
 
-func getTypesFromPlan(p *sbmodel.ServicePlan) (*model.Types, error) {
+func getTypesFromPlan(p *brokerapi.ServicePlan) (*brokerapi.Types, error) {
 	err := fmt.Errorf("Did not find usable types for plan %s", p.ID)
 
 	if p.Metadata == nil {
 		return nil, err
 	}
-	if _, ok := p.Metadata.(map[string]interface{})[model.INSTANCE_TYPE]; !ok {
+	if _, ok := p.Metadata.(map[string]interface{})[brokerapi.InstanceType]; !ok {
```
Comment (arschles) [inline contrib/pkg/broker/k8s/controller/controller.go]: can you check this and the below type assertions?

File: pkg/apis/componentconfig/types.go (hunk 5 above)
Comment (arschles) [inline pkg/apis/componentconfig/types.go]: We should add the above explanation as a package-level Godoc comment

File: pkg/controller/controller.go [hunk 9]
```diff
@@ -14,65 +14,92 @@ See the License for the specific language governing permissions and
...
+
+func createController(h Handler, w *watch.Watcher) *controller {
+	return &controller{
+		handler: h,
+		watcher: w,
+	}
+}
+
+func (c *controller) Run() error {
+	if c.watcher == nil {
```
Comment (arschles) [inline pkg/controller/controller.go]: Is this supposed to be concurrency-safe? I'm asking from curiosity, not blocking for this PR

**Comment by krancour** [comment]
@arschles and I discussed offline and have agreed that the nits he picked come from upstream and if we want to address them, we can do so later.

**Comment by arschles** [comment]
Follow up to @krancour's comment: I'll go through this PR and submit issues to the queue for follow-ups that I think are worthwhile.

**Comment by arschles** [comment]
LGTM3, pending CI

**Comment by MHBauer** [comment]
not sure I understand 
pkg/cmd/server/server.go → cmd/apiserver/app/server/server.go

why move it here? @krancour @pmorie ?

**Comment by krancour** [comment]
@MHBauer I moved everything to where it seemed like it belonged based on precedents found upstream in k8s. It's possible I made mistakes here and there. If you think I got it wrong, I'm not offended if you want to fix it.

**Comment by krancour** [comment]
So, I based it off this: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go

I think this was a tiny mistake and that the file needs to move up one level.
